### PR TITLE
[SYCL] Add FPGA kernel_attribute properties

### PR DESCRIPTION
**Pull request by mendell27**
Support streaming_interface, register_map_interface, pipelined.

Only partial support for register_map_interface.

### CODE REVIEW

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVBase.td [hunk 1]
```diff
@@ -3098,6 +3098,11 @@ def SPIRV_EM_SchedulerTargetFmaxMhzINTEL      : I32EnumAttrCase<"SchedulerTarget
     Capability<[SPIRV_C_FPGAKernelAttributesINTEL]>
   ];
 }
+def SPIRV_EM_StreamingInterfaceINTEL          : I32EnumAttrCase<"StreamingInterfaceINTEL", 6154> {
+  list<Availability> availability = [
+    Capability<[SPIRV_C_FPGAKernelAttributesINTEL]>
+  ];
+}
```
Comment (bader) [inline mlir/include/mlir/Dialect/SPIRV/IR/SPIRVBase.td:3105]: I don't think this should be committed.
If you intend to extend SPIR-V dialect, please, commit this change to llvm.org project directly (or sycl-mlir branch).

Comment (mendell27) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVBase.td:3105]: Okay.  I guess I am a bit unclear about exactly what goes where.  I can pull this file out of this PR and add it LLVM directly.  If I do so, are there any other files that should be updated too?

Comment (mendell27) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVBase.td]: Removed and added in [D140476](https://reviews.llvm.org/D140476)

**Review by dm-vodopyanov — CHANGES_REQUESTED** [review]
@mendell27 could you please share a link to the spec? The only one I found is [sycl_ext_intel_fpga_kernel_interface_properties.asciidoc](https://github.com/intel/llvm/blob/b2d6fdfd63fe8dd7be98e113cde9bc7c3d9a21d8/sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc) and the implementation in this PR doesn't match this spec.

**Comment by mendell27** [comment]
> @mendell27 could you please share a link to the spec? The only one I found is [sycl_ext_intel_fpga_kernel_interface_properties.asciidoc](https://github.com/intel/llvm/blob/b2d6fdfd63fe8dd7be98e113cde9bc7c3d9a21d8/sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc) and the implementation in this PR doesn't match this spec.

That is the one that I used.  Where does it differ?  The bottom of the file is the implementation details.  Do I have to move those to another file?

**Comment by dm-vodopyanov** [comment]
> > @mendell27 could you please share a link to the spec? The only one I found is [sycl_ext_intel_fpga_kernel_interface_properties.asciidoc](https://github.com/intel/llvm/blob/b2d6fdfd63fe8dd7be98e113cde9bc7c3d9a21d8/sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc) and the implementation in this PR doesn't match this spec.
> 
> That is the one that I used. Where does it differ? The bottom of the file is the implementation details. Do I have to move those to another file?

The implementation introduces new API in `sycl::ext::oneapi::experimental` namespace but spec says this API must be in `sycl::ext::intel::experimental`. Either implementation or spec should be fixed.

**Comment by mendell27** [comment]
I have moved the include file to sycl/ext/intel/experimental/fpga_kernel_attributes.hpp

File: llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp [hunk 2]
```diff
@@ -202,6 +206,26 @@ attributeToExecModeMetadata(Module &M, const Attribute &Attr) {
                                             MDNode::get(Ctx, MD));
   }
 
+  auto getIpInterface = [](const char *name, LLVMContext &Ctx,
```
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp:209]:
```suggestion
  auto getIpInterface = [](const char *Name, LLVMContext &Ctx,
```

**Review by AlexeySachkov — CHANGES_REQUESTED** [review]

File: llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp (hunk 2 above)
Comment (mendell27) [reply llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp]: Changed

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
+//==----- fpga_kernel_attributes.hpp - SYCL properties associated with FPGA
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:1]: According to spec,

> an implementation supporting this extension must predefine the macro SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES

Please add this macro to https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/feature_test.hpp.in

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+template <int Value>
+struct PropertyMetaInfo<intel::experimental::pipelined_key::value_t<Value>> {
+  static constexpr const char *name = "sycl-pipelined";
+  static constexpr int value = Value;
+};
+
+} // namespace detail
+} // namespace ext::oneapi::experimental
+} // __SYCL_INLINE_VER_NAMESPACE(_V1)
+} // namespace sycl
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:154]: Is it possible to add some tests for this new API?

**Review by dm-vodopyanov — CHANGES_REQUESTED** [review]

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
+//==----- fpga_kernel_attributes.hpp - SYCL properties associated with FPGA
+// kernel attributes ---==//
```
Comment (dm-vodopyanov) [inline sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:2]: The extension has name sycl_ext_intel_fpga_kernel_interface_properties. Better to rename the file from fpga_kernel_attributes.hpp to fpga_kernel_interface_properties.hpp

**Review by AlexeySachkov — APPROVED** [review]
`sycl-post-link` part LGTM

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp (hunk 4 above)
Comment (mendell27) [reply sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:154]: added one.  Is that enough?

File: sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp (hunk 5 above)
Comment (mendell27) [reply sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:2]: added

Comment (dm-vodopyanov) [reply sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:2]: Sorry, looks like this wasn't applied

Comment (mendell27) [reply sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp:2]: You are right.  The comment was in the wrong place.  I will change the name in the next push, but I am just struggling with ensuring my test is working after the file rename

Comment (mendell27) [reply sycl/include/sycl/ext/intel/experimental/fpga_kernel_attributes.hpp]: **now updated**

**Review by dm-vodopyanov — APPROVED** [review]

**Pull request merged**